internal/repository/master_index.go

```diff
@@ -237,24 +234,35 @@ func (mi *MasterIndex) RebuildIndex(packBlacklist restic.IDSet) (newIndex *Index
 			newIndex.Store(pb)
 		}
 
-		if !idx.Final() {
-			debug.Log("index %d isn't final, don't add to supersedes field", i)
-			continue
-		}
+		if idx.Final() {
+			id, err := idx.ID()
+			if err != nil {
+				debug.Log("index %d does not have an ID: %v", err)
+				return nil, err
+			}
 
-		id, err := idx.ID()
-		if err != nil {
-			debug.Log("index %d does not have an ID: %v", err)
-			return nil, nil, err
-		}
+			debug.Log("adding index id %v to supersedes field", id)
 
-		debug.Log("adding index id %v to supersedes field", id)
+			err = newIndex.AddToSupersedes(id)
+			if err != nil {
+				return nil, err
+			}
+			obsolete.Insert(id)
+		} else {
+			debug.Log("index %d isn't final, don't add to supersedes field", i)
+		}
 
-		err = newIndex.AddToSupersedes(id)
-		if err != nil {
-			return nil, nil, err
+		if IndexFull(newIndex) {
```

Comment:
The code in `internal/index/index.go` uses a much higher threshold of 3000 packs before creating a new index. `IndexFull` creates much smaller index files with only around 2000 blobs in them. I think we want the larger limit of 3000 packs (which seems to generate 10-15MB index files for my large test repository). Or maybe something like 100k blobs in a single index.

Reply:
You are right. The `IndexFull` is also called by `restic backup` and additionally to the number of blobs checks. I just checked it right now and saw there is also a check that an index which is not older than 2 Minutes is not declared as "Full". So I expect that also here the index will get really big (as it is "young" and will be filled really fast from memory) before it gets saved.

However, I do not see why `prune` should have a different threshold than `backup`. I suggest to keep this and change the default values (and the logic) of `IndexFull` in a separate PR.

Reply:
It should probably be something like "full if too old or large enough". And then use e.g. 10 minutes and 50k blobs as limits. At least for my backups (mostly some small files with changes) currently the intermediate indexes are uploaded every few minutes, that timespan would increase to 10 or 15 minutes.

Reply:
I made a new PR #2749 to change this.